compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+
+        Runtime.getRuntime.exec(fullArgs).waitFor() == 0
+      } else true
+
+
+      // First we try to compile the java files in the directory:
+      val javaFiles = files.filter(_.getName.endsWith(".java")).map(_.getAbsolutePath)
+      val javaCompiledBefore = compileWithJavac(javaFiles)
+
+      // Then we compile the scala files:
+      val reporter = TestReporter.parallelReporter(this, logLevel =
+        if (suppressErrors || suppressAllOutput) ERROR + 1 else ERROR)
+      val driver =
+        if (times == 1) new Driver { def newCompiler(implicit ctx: Context) = new Compiler }
+        else new Driver {
+          def newCompiler(implicit ctx: Context) = new Compiler
+
+          private def ntimes(n: Int)(op: Int => Reporter): Reporter =
+            (emptyReporter /: (1 to n)) ((_, i) => op(i))
+
+          override def doCompile(comp: Compiler, files: List[String])(implicit ctx: Context) =
+            ntimes(times) { run =>
+              val start = System.nanoTime()
+              val rep = super.doCompile(comp, files)
+              ctx.echo(s"\ntime run $run: ${(System.nanoTime - start) / 1000000}ms")
+              rep
+            }
+        }
+
+      val allArgs = addOutDir(flags)
+      driver.process(allArgs ++ files.map(_.getAbsolutePath), reporter = reporter)
+
+      // If the java files failed compilation before, we try again after:
+      if (!javaCompiledBefore)
+        assert(compileWithJavac(javaFiles), s"java compilation failed for ${javaFiles.mkString(", ")}")
+
+      reporter
+    }
+
+    private[ParallelTesting] def execute(): this.type = {
+      assert(_targetsCompiled == 0, "not allowed to re-use a `CompileRun`")
+
+      if (allTargets.nonEmpty) {
+        val pool = threadLimit match {
+          case Some(i) => JExecutors.newWorkStealingPool(i)
+          case None => JExecutors.newWorkStealingPool()
+        }
+
+        if (interactive && !suppressAllOutput) pool.submit(statusRunner)
+
+        allTargets.foreach { target =>
+          pool.submit(compilationRunnable(target))
+        }
+
+        pool.shutdown()
+        if (!pool.awaitTermination(10, TimeUnit.MINUTES))
+          throw new TimeoutException("Compiling targets timed out")
+
+        if (didFail) {
+          echo {
+            """|
+               |================================================================================
+               |Test Report
+               |================================================================================
+               |Failing tests:""".stripMargin
+          }
+          failedCompilationTargets.toArray.sorted.foreach(echo)
+          failureInstructions.iterator.foreach(echo)
+        }
+      }
+      else echo {
+        regex
+          .map(r => s"""No files matched regex "$r" in test""")
+          .getOrElse("No tests available under target - erroneous test?")
+      }
+
+      this
+    }
+  }
+
+  @inline private final def isCompilable(f: JFile): Boolean = {
```

Comment:
Why not? The heuristics of 2.11 (and 2.12) are satisfied AFAICT and in dotty it will be enforced.